Publicize: Sync WordPress.com and Jetpack UI

mdawaffe opened the pull request:
#### Changes proposed in this Pull Request:

* New (and already deprecated) filter for customizing Publicize Settings URL. Do not use :) It will go away.
* New (and already deprecated) filter for customizing Publicize Settings Page. Also do not use :) Same.
* Code formatting tweaks
* Translation fixes (mostly `/* translators: */` comments).
* Improve default message handling: Show as placeholder if textarea is empty
* Improve custom message handling: On page load, if there is a "non-trivial" custom message (one that is not empty and does not match the default message), open the UI so that the author remembers. (Use case: when a custom message is explicitly set to the default message, then the post title changes.)

#### Testing instructions:

##### Does it Work?
1. Add at least two Publicize connections on a test site.
2. Create a post in the Classic (non-Gutenberg) editor.
3. Set it to publicize to one connection, but not the others.
4. Save as Draft.
5. See that the Publicize settings are correct after saving the draft (set to publicize to one connection, but not the others).
6. Publish the post.
7. See the publicize results on the one connection (e.g., the tweet).
8. In the post editor, see all Publicize checkboxes disabled (this is the same as the old behavior).

##### Publicize Message UI
1. In the Classic (non-Gutenberg) editor, create a new post.
2. Set the title to "Hello World".
3. Set the contents to "Yay".
4. Save as Draft.
5. See the Publicize settings are still **closed** on the new page load after saving the draft.
6. Edit the Publicize message. Erase all contents in the message.
7. See the message box's placeholder matches the post's title.
8. Save as draft.
9. See the Publicize message UI is still **closed** on the new page load after saving the draft.
10. Edit the Publicize message. Set the Publicize message to "Something else".
11. Save as draft.
12. See the Publicize message UI is now **open** on the new page load after saving the draft.
13. Edit the Publicize settings. Set the Publicize message to "Hello World".
14. Save as draft.
15. See the Publicize message UI is again **closed** on the new page load after saving the 
draft.
16. Set the title to "New Title"
17. Save as Draft.
18. See the Publicize message UI is again **open** on the new page load after saving the 
draft.

#### Proposed changelog entry for your changes:

Updated Jetpack's and WordPress.com's Publicize user interface to match.

mdawaffe committed:
Publicize: Sync with WordPress.com

mdawaffe committed:
Publicize: Fix translation bugs

mdawaffe committed:
Publicize: Sync WordPress.com's custom Settings URL.

This will go away in a future version of Jetpack.

mdawaffe committed:
Publicize: Sync WordPress.com's custom Settings Page.

This will go away in a future version of Jetpack.

mdawaffe committed:
Publicize: Show Custom Message UI if Set

If a Custom Message has been set (on an earlier page load), open the UI on
page load so that the author doesn't forget.

Makes it easier to see that the Custom Message may need updating if, for
example, the post title changes.

mdawaffe committed:
Publicize: Keep Track of Active Service Doneness

In addition to looking at the `$all_done` flag, keep track of whether all
active services are done.

gravityrail commented:
Did you see this [other PR](https://github.com/Automattic/jetpack/pull/10041) where @Tug brought across the publicize connect button from Calypso? It's a prototype of an unrelated feature so unlikely to be merged before this release.

Might be of interest if we decide to move the wp-admin UI into the Jetpack Settings area.

mdawaffe committed:
Publicize: Clarify API

Part of unification of WP.com and Jetpack Publicize code.

Tug commented:
Just a note that it's still a prototype and it's super limited at the moment:
- it's hacky
- it doesn't work if publicize is disabled
- it needs a page refresh if user enables publicize from the settings
- etc.
The approach taken here looks way more solid 👍

ockham commented:
Rebased.

kraftbj commented:
We'll need to do some tender loving care for the WP.com. Since Fusion isn't applying the diff cleanly, I'm going to mark this as okay for the wpcom test.

ockham commented on modules/publicize/publicize-jetpack.php:
```diff
@@ -231,20 +238,35 @@ function display_disconnected() {
+
+		if ( ! $xml->isError() ) {
+			$response = $xml->getResponse();
+			$this->receive_updated_publicize_connections( $response );
```
Since both methods have almost identical implementations, can we merge them and make the `'globalize'`/`'unglobalize'` param to `$xml->query` an argument?

mdawaffe replied:
We could - despite the number of red and green lines, I was going for minimal changes :) If you feel strongly about it, let's get it in. Otherwise, I think it can be a different PR.

ockham commented on modules/publicize/publicize.php:
```diff
@@ -245,33 +407,146 @@ function show_options_popup( $service_name, $connection ) {
+		if ( $this->is_connecting_connection( $connection ) ) {
+			return true;
+		}
+		$connection_data = $connection->get_meta( 'connection_data' );
```
Will this work in the Jetpack case (`$connection` being an array there)?

mdawaffe replied:
No - it definitely won't :) Thanks!

mdawaffe replied:
Fixed in 5087018f6d0670ffd6c043bc96902e4bbdffeb62.

ockham commented on modules/publicize/publicize.php:
```diff
@@ -245,33 +407,146 @@ function show_options_popup( $service_name, $connection ) {
+	 * @return bool
+	 */
+	function is_connecting_connection( $connection ) {
+		$connection_data = $connection->get_meta( 'connection_data' );
```
Same as https://github.com/Automattic/jetpack/pull/10043/files#r226724541

ockham commented on modules/publicize/publicize.php:
```diff
@@ -245,33 +407,146 @@ function show_options_popup( $service_name, $connection ) {
+	 * @param object|array The Connection object (WordPress.com) or array (Jetpack)
+	 * @return bool
+	 */
+	function is_valid_facebook_connection( $connection ) {
```
Could absorb the check whether we're dealing with a Facebook connection, and thus generalized to `is_valid_connection()` (to possibly hold other checks later).

mdawaffe replied:
That makes sense. I guess let's wait to change it when we have that need, though?

ockham commented on modules/publicize/publicize.php:
```diff
@@ -245,33 +407,146 @@ function show_options_popup( $service_name, $connection ) {
+
+/*
+ * Site Data
+ */
```
Indented like this on purpose?

mdawaffe replied:
Yes - the mark "sections" in the big all-purpose class. They stand out, don't they? :)

ockham commented on modules/publicize/publicize.php:
```diff
@@ -245,33 +407,146 @@ function show_options_popup( $service_name, $connection ) {
+
+/*
+ * Posts
+ */
```
Purposefully indented like this?

mdawaffe replied:
[Yes](https://github.com/Automattic/jetpack/pull/10043#discussion_r227519433).

ockham commented on modules/publicize/ui.php:
```diff
@@ -22,6 +24,14 @@ function __construct() {
+		$this->publicize_settings_url = apply_filters_deprecated(
+			'jetpack_override_publicize_settings_url',
+			array( admin_url( 'options-general.php?page=sharing' ) ),
+			'6.5',
```
Requires bumping I guess

ockham commented on modules/publicize/ui.php:
```diff
@@ -57,7 +69,7 @@ function sharing_menu() {
 			'publish_posts',
 			'sharing',
-			array( &$this, 'wrapper_admin_page' )
+			array( $this, 'wrapper_admin_page' )
```
I think I missed another `&$this` on L77 when rebasing

```php
Jetpack_Admin_Page::wrap_ui( array( &$this, 'management_page' ), array( 'is-wide' => true ) );
```

ockham commented on modules/publicize/ui.php:
```diff
@@ -131,10 +144,22 @@ public static function denied_notice() { ?>
+		$override_publicize_settings_page = apply_filters_deprecated(
+			'jetpack_override_publicize_settings_page',
+			array( false ),
+			'6.5',
```
Needs bumping

ockham commented on modules/publicize/ui.php:
```diff
@@ -637,13 +693,19 @@ private function get_metabox_form_connected( $services ) {
+
+					$unique_id = $this->publicize->get_connection_unique_id( $connection );
+
+					// Was this connections (OR, old-format service) already Publicized to?
```
Typo, `connections` (should be singular)

ockham reviewed (commented):
Looking good -- love the PHPdoc comments, so helpful, especially for someone pretty much unfamiliar with this codebase!

I'll give this some testing on Monday. Testing instructions appreciated :) Otherwise, I guess I just need to test this PR with Jetpack, and the Fusion-generated phab diff on WP.com...?

mdawaffe committed:
Publicize: Typos

Props @ockham

mdawaffe committed:
Publicize: Bump deprecated filters version.

Props @ockham

mdawaffe committed:
Publicize: Code standards

Props @ockham

mdawaffe committed:
Publicize: Variable name consistency

Fixes PHP Notice:

> Undefined variable: service in modules/publicize/publicize.php on line 154

mdawaffe committed:
Publicize: Use the `get_connection_meta()` abstraction

Fixes PHP Fatal:
> Uncaught Error: Call to a member function get_meta() on array in modules/publicize/publicize.php:434

mdawaffe commented:
I cleaned up a bunch of things based on your feedback, @ockham, including a Fatal Error 😱. Thanks!

ockham reviewed (dismissed):
Thanks @mdawaffe! LGTM :+1:

mdawaffe commented:
> **Caution: This PR has changes that must be merged to WordPress.com**
> Please review this diff before merging: D19814-code. _(newly created revision)_

Not sure if we can fix the bot, but the correct diff is D17427-code.

ockham commented on modules/publicize/publicize.php:
```diff
@@ -107,33 +107,210 @@ function __construct() {
-	abstract function get_connection_id( $connection );
+	 * Get services for the given blog and user.
+	 *
+	 * Can return all avaialable services or just the ones with an active connection.
```
Typo, `avaialable` (missed it at first, only noticed now when looking at the WP.com side diff)

ockham commented on modules/publicize/publicize.php:
```diff
@@ -204,23 +376,13 @@ function get_display_name( $service_name, $connection ) {
-	}
-
+	/**
+	 * Weather the user needs to select additional options after connecting
```
Should be `Whether` :)

mdawaffe replied:
Haha :)

mdawaffe committed:
Publicize: Typos

Props @ockham

kraftbj commented:
I'm not sure about getting the bot to ack the other diff, but we can manually force the wpcom test to pass. I did so now, but I believe we'll have to redo it if there are future commits before it can merge.

mdawaffe committed:
Publicize: Make `jetpack_override_publicize_settings_page` filter easier to use.

It's deprecated, but let's still make it simpler.

Props @ockham.

ockham reviewed (dismissed):
Went through testing instructions, works like a charm! :shipit:

kraftbj commented:
Seeing a warning on JN: ```
Notice: Trying to get property of non-object in /srv/public/wp-content/plugins/jetpack-dev/modules/publicize/ui.php on line 735```

kraftbj reviewed (dismissed):
Seeing a warning on JN: Notice: Trying to get property of non-object in /srv/public/wp-content/plugins/jetpack-dev/modules/publicize/ui.php on line 735

mdawaffe committed:
Publicize: Simplify meta lookup logic

Connections are arrays in Jetpack and objects in WordPress.com.

* When looking for their corresponding services' names, just look at the
  service itself instead of the connection's copy.
* Ensure we're accessing their data via helper methods (like
  `->get_connection_meta()`).

mdawaffe commented:
> Seeing a warning on JN: ` Notice: Trying to get property of non-object in /srv/public/wp-content/plugins/jetpack-dev/modules/publicize/ui.php on line 735`

@kraftbj, Thanks! This should be fixed as of ea6e6b639d6eeacaf16b3bcb10bff3c4b203b173. Could you take another look?

oskosk reviewed (approved):
Tested perfectly with provided instructions! :shipit:

kraftbj commented:
Sync'd to wp.com via r182397-wpcom and r182406-wpcom
